_posts/2019-01-25-building-apps-with-plug-router.md

```diff
@@ -0,0 +1,345 @@
+
+### Rendering templates
+
+Elixir includes EEx, Embedded Elixir, which lets us parse strings and files and evaluate the embedded Elixir.
+For our purposes we'll focus on `EEx.eval_file/2` and use it as the basis for our own `render/3` function.
+
+To keep our `render/3` function familiar we'll pass in our connection struct, a template without the ".eex" extension, and any variable bindings we may want to provide to the embedded Elixir.
+We want a function invocation that will look something like this:
+
+```elixir
+render(conn, "index.html", portfolio: [])
+```
+
+Now that we know what we want it's time to build it!
+
+The `EEx.eval_file/2` function takes our template's filepath along with variable bindings and returns the computed string, our response body.
+With EEx doing all the heavy lifting our `render/3` function needs to do little more than build the complete filepath and send along the response with `send_resp/3`:
+
+```elixir
+@template_dir "lib/router_example/templates"
+
+...
+
+defp render(%{status: status} = conn, template, assigns \\ []) do
+  body =
+    @template_dir
+    |> Path.join(template)
+    |> String.replace_suffix(".html", ".html.eex")
+    |> EEx.eval_file(assigns)
+
+  send_resp(conn, (status || 200), body)
+end
+```
+
+We've set it up for EEx to look for our templates in the directory `lib/router_example/templates` so let's create that directory.
+Next we'll create two templates `index.html.eex` and `contact.html.eex`.
+
+You can find the code for `index.html.eex` [here]() and `contact.html.eex` [here](), we won't focus on HTML or CSS today.
+
+### Sending and recieving JSON
+
+Building JSON endpoints in Plug.Router is less much less work than the template rendering we just covered.
+
+The first thing we need is a library to parse and encode JSON, for our purpose we'll use [jason]():
+
+```elixir
+{:jason, "~> 1.1"}
+```
+
+Now would be a good time to run `mix deps.get` before we forget.
+Once we've done that we can move on to updatin our router to handle incoming JSON via the `Plug.Parsers` plug.
+Let's open `lib/router_example/router.ex` and update our plugs to include `Plug.Parsers` with `Jason` as our decoder:
+
+```elixir
+plug Plug.Parsers, parsers: [:json],
+                   pass: ["text/*"],
+                   json_decoder: Jason
+plug :match
+plug :dispatch
+```
+
+That's _really_ all we need to do for JSON.
+If we want to keep things simple, we can leverage `Jason.encode/1` or `Jason.encode!/1` along with `send_resp/3` and be done:
+
+```elixir
+{:ok, json} = Jason.encode(result)
+send_resp(conn, 200, json)
+```
+
+Or if we want a little more polish we could make a `render_json/2`:
+
+```elixir
+defp render_json(%{status: status} = conn, data) do
+  body = Jason.encode!(data)
+  send_resp(conn, (status || 200), body)
+end
+```
+
+For the remainder of the post we'll use the `render_json/2` approach.
```

Comment:
I think it might be worth noting that there is additional work needed if you wanna meet a standard like JSONAPI. I figure a lot of folks looking for a solution like this may be having to adhere to some stricter rules like that but might not need all the things Phoenix offers.